Analysis/SLDCorrection/src/FindParticle.cc

```diff
@@ -0,0 +1,238 @@
+			}
+			for ( unsigned int i_par = 0 ; i_par < truePVANeutralDecayProducts.size() ; ++i_par )
+			{
+				if ( linkedNeutralPFO == truePVANeutralDecayProducts[ i_par ] ) recoParticleExist = true;
+			}
+			if ( !recoParticleExist && linkedNeutralPFO != linkedRecoLepton )
+			{
+				if ( ( linkedNeutralPFO->getTracks() ).size() == 0 )
+				{
+					truePVANeutralDecayProducts.push_back( linkedNeutralPFO );
+				}
+				else
+				{
+					truePVAChargedDecayProducts.push_back( linkedNeutralPFO );
+				}
+			}
+		}
+		else
+		{
+			getTruePVADecayProducts( daughter , SLDLepton , trueNeutrino , linkedRecoLepton , weightChargedPFOtoMCP , weightChargedMCPtoPFO , truePVANeutralDecayProducts , truePVAChargedDecayProducts , RecoMCParticleNav , MCParticleRecoNav );
+		}
+	}
+}
+
+void getLinkedRecoLepton(	EVENT::MCParticle *SLDLepton , pfo &linkedRecoLepton , 
+							LCRelationNavigator RecoMCParticleNav , LCRelationNavigator MCParticleRecoNav , 
+							float &weightRecoLeptoMCLep , float &weightMCLeptoRecoLep )
+{
+	ReconstructedParticle* linkedRecoLep{};
+	bool foundLinkedRecoLepton = false;
+	linkedRecoLep = getLinkedPFO( SLDLepton , RecoMCParticleNav , MCParticleRecoNav , true , false , weightRecoLeptoMCLep , weightMCLeptoRecoLep , foundLinkedRecoLepton );
+	if ( !foundLinkedRecoLepton )
+	{
+		for ( long unsigned int i_daughter = 0 ; i_daughter < ( SLDLepton->getDaughters() ).size() ; ++i_daughter )
+		{
+			EVENT::MCParticle *daughter = SLDLepton->getDaughters()[ i_daughter ];
+			getLinkedRecoLepton( daughter , linkedRecoLep , RecoMCParticleNav , MCParticleRecoNav , weightRecoLeptoMCLep , weightMCLeptoRecoLep );
+		}
+	}
+	linkedRecoLepton = linkedRecoLep;
+}
+
+EVENT::ReconstructedParticle* getLinkedPFO( EVENT::MCParticle *mcParticle , LCRelationNavigator RecoMCParticleNav , LCRelationNavigator MCParticleRecoNav , bool getChargedPFO , bool getNeutralPFO , float &weightPFOtoMCP , float &weightMCPtoPFO , bool &foundlinkedPFO )
+{
+	streamlog_out(DEBUG1) << "" << std::endl;
+	streamlog_out(DEBUG1) << "	Look for PFO linked to visible MCParticle:" << std::endl;
+	streamlog_out(DEBUG1) << *mcParticle << std::endl;
+	ReconstructedParticle* linkedPFO{};
+	foundlinkedPFO = false;
+	const EVENT::LCObjectVec& PFOvec = MCParticleRecoNav.getRelatedToObjects( mcParticle );
+	const EVENT::FloatVec&  PFOweightvec = MCParticleRecoNav.getRelatedToWeights( mcParticle );
+	streamlog_out(DEBUG0) << "	Visible MCParticle is linked to " << PFOvec.size() << " PFO(s)" << std::endl;
+	weightPFOtoMCP = 0.0;
+	weightMCPtoPFO = 0.0;
+	double maxweightPFOtoMCP = 0.;
+	double maxweightMCPtoPFO = 0.;
+	int iPFOtoMCPmax = -1;
+	int iMCPtoPFOmax = -1;
+	for ( unsigned int i_pfo = 0; i_pfo < PFOvec.size(); i_pfo++ )
+	{
+		double pfo_weight = 0.0;
+		double trackWeight = ( int( PFOweightvec.at( i_pfo ) ) % 10000 ) / 1000.0;
+		double clusterWeight = ( int( PFOweightvec.at( i_pfo ) ) / 10000 ) / 1000.0;
```

Comment:
You could use the [new functionality in MarlinUtil](https://github.com/iLCSoft/MarlinUtil/blob/13d75c07698d28564aa6366c61dc6c0b20d58769/source/include/MarlinUtil.h#L64-L80) for this.